apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,835 @@
+}
+
+// Listener embodies the concept of a logical endpoint where a Gateway can
+// accept network connections. Each listener in a Gateway must have a unique
+// combination of Hostname, Port, and Protocol. This will be enforced by a
+// validating webhook.
+type Listener struct {
+	// Name is the name of the Listener. If more than one Listener is present
+	// each Listener MUST specify a name. The names of Listeners MUST be unique
+	// within a Gateway.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:MaxLength=253
+	Name string `json:"name"`
+
+	// Hostname specifies the virtual hostname to match for protocol types that
+	// define this concept. When unspecified, "", or `*`, all hostnames are
+	// matched. This field can be omitted for protocols that don't require
+	// hostname based matching.
+	//
+	// For HTTPRoute objects, there is an interaction with the
+	// `spec.hostnames` array. When both listener and route specify hostnames,
+	// there must be an intersection between the values for a Route to be admitted.
+	// For example, a Gateway with `*.example.com` would admit a Route that included
+	// `foo.example.com` as a hostname, but not a Route that *only* included
+	// `foo.acme.io` as a hostname. A Route that included both `foo.example.com`
+	// and `foo.acme.io` would be admitted, but the `foo.acme.io` hostname would
+	// be silently ignored.
+	//
+	// Support: Core
+	//
+	// +optional
+	Hostname *Hostname `json:"hostname,omitempty"`
+
+	// Port is the network port. Multiple listeners may use the
+	// same port, subject to the Listener compatibility rules.
+	//
+	// Support: Core
+	Port PortNumber `json:"port"`
+
+	// Protocol specifies the network protocol this listener expects to receive.
+	// The GatewayClass MUST apply the Hostname match appropriately for each
+	// protocol:
+	//
+	// * For the "TLS" protocol, the Hostname match MUST be
+	//   applied to the [SNI](https://tools.ietf.org/html/rfc6066#section-3)
+	//   server name offered by the client.
+	// * For the "HTTP" protocol, the Hostname match MUST be
+	//   applied to the host portion of the
+	//   [effective request URI](https://tools.ietf.org/html/rfc7230#section-5.5)
+	//   or the [:authority pseudo-header](https://tools.ietf.org/html/rfc7540#section-8.1.2.3)
+	// * For the "HTTPS" protocol, the Hostname match MUST be
+	//   applied at both the TLS and HTTP protocol layers.
+	//
+	// Support: Core
+	Protocol ProtocolType `json:"protocol"`
+
+	// TLS is the TLS configuration for the Listener. This field
+	// is required if the Protocol field is "HTTPS" or "TLS" and
+	// ignored otherwise.
+	//
+	// The association of SNIs to Certificate defined in GatewayTLSConfig is
+	// defined based on the Hostname field for this listener.
+	//
+	// The GatewayClass MUST use the longest matching SNI out of all
+	// available certificates for any TLS handshake.
+	//
+	// Support: Core
+	//
+	// +optional
+	TLS *GatewayTLSConfig `json:"tls,omitempty"`
```

Comment:
Chances are all listeners will have the same TLS config. Or at least a few of them share it. And the configuration is rotated to together.
Here is a proposal:
1. TLS config should be a type of its own. Referenced from listener.
2. Add a status field to indicate which TLS config is currently active. I am think this is an error prone areas and debugging TLS without knowing which config is active requires much more efforts than just looking at `status` fields.

Reply:
We need to do a better job at communicating this, but our idea here is that each listener would represent a unique combination of protocol, port, and hostname. In most cases we've seen/expected this to translate into something like this:

```yaml
- hostname: foo.com
  port: 80  
  protocol: HTTP
- hostname: foo.com
  port: 443  
  protocol: HTTPS
  tlsConfig:
    // foo.com cert ref
- hostname: bar.com
  port: 80  
  protocol: HTTP
- hostname: bar.com
  port: 443  
  protocol: HTTPS
  tlsConfig:
    // bar.com cert ref
``` 

Is there a use case you're thinking where it would be helpful for listeners to share the same cert?

Reply:
Also this is configuring TLS for a listener not a for a gateway. The struct name is a bit confusing.

Reply:
I like this. Thank you for clarification.